### ISSUE TITLE
css-vars and shadow dom

### ISSUE THREAD

**Issue by dvelasquez** [jhildenbiddle/css-vars-ponyfill#24]
Hi guys, i'm trying to use cssVars() in a custom element v1 in internet explorer 11. So far i can't make it work with shadow dom (throught a polyfill).

This is a link to a repo where you can test it.

https://github.com/dvelasquez/ponyfill-issue-ie11

Let me know if you need anything else

**Comment by jhildenbiddle** [comment]
Hi @dvelasquez.

The ponyfill currently does not handle Shadow DOM elements.

I did take a look at your demo repo, and the good news is that I think there are some enhancements that can be made to the ponyfill to handle your specific scenario.

First things first though. I noticed the following in you [src/index.html](https://github.com/dvelasquez/ponyfill-issue-ie11/blob/8cb38d68e50d7b86b8522a148997c67f24ae3774/src/index.html#L11-L13) file:

```css
html {
  --test-color: blue;
}
```

The ponyfill requires CSS custom properties to be defined within a `:root` selector:

```css
:root {
  --test-color: blue;
}
```

This won't address your issue, but I wanted to point it for future use.

Regarding your custom element issue, the following two enhancements would allow the ponyfill to handle Shadow DOM `<style>` elements:

1. Add a third argument to the [`options.onComplete`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsoncomplete) callback: an object containing all custom property name:value pairs.

   ```javascript
   cssVars({
     onComplete(cssText, styleNode, variablesObject) {
       // ...
     }
   });
   ```

1. Add an option to specify the root element to parse and append styles to. Currently the ponyfill assumes the root element is `document`, but this new option would allow devs to specify a `shadowRoot` element.

   ```javascript
   cssVars({
     rootElement: el.shadowRoot
   });
   ```

The combination of these two updates would allow the ponyfill to collect CSS custom properties from the standard DOM, then pass them to a subsequent ponyfill call (via [`options.variables`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsvariables)) that processes Shadow DOM `<style>` elements by setting the `rootElement` option to a web component's `shadowRoot`:

```javascript
cssVars({
  onComplete(cssText, styleNode, variablesObject) {
    const elms = document.body.querySelector('*');

    for (let i = 0, elm; elm = elms[i]; ++i) {
      if (elm.shadowRoot) {
        cssVars({
          rootElement: elm.shadowRoot,
          variables: variablesObject
        });
      }
    }
  }
});
```

The above code would handle *all* custom elements. Using your demo repo as an example, the end result would be something like this:

```html
<test-shadow>
  #shadow-root
    <style>
      .shadowDomTest {
        color: var(--test-color);
      }
    </style>
    <style id="css-vars-ponyfill">
      .shadowDomTest {
        color: blue;
      }
    </style>
    <div>
      <h1 class="shadowDomTest"><slot name="content"></slot></h1>
    </div>
    <span slot="content">this text should be blue</span>
</test-shadow>
```

Ideally, this would all be baked into the ponyfill, able to be triggered by setting an option named something like `processShadowDOM`:

```javascript
cssVars({
  processShadowDOM: true  // default: false
});
```

I can add this to the backlog, but I'm not sure when I will get around to implementing this. It will take some time to make the changes and write tests, especially considering that I am not very familiar with web components.

**Comment by jhildenbiddle** [comment]
Hi again, @dvelasquez.

The changes outlined above have been made and released as v1.10.0. You can review the  [CHANGELOG.md](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/CHANGELOG.md) for details.

To address your issue, set the new [`shadowDOM`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsshadowdom) option to `true`:

```javascript
cssVars({
  shadowDOM: true
});
```

This new feature works on modern browsers with native shadow DOM support and older browsers with the appropriate polyfills (tested with IE11, Chrome 26, Edge 13, Firefox 30, and Safari 8).

Hopefully this resolves your issue. If not, or if you have any questions, let me know. Thx!

**Comment by dvelasquez** [comment]
Thanks very much, i've made the changes and now is working.

Again, thanks @jhildenbiddle for your amazing work!

**Comment by jhildenbiddle** [comment]
Happy to help, @dvelasquez. Glad everything works as expected. 😄

**Comment by dvelasquez** [comment]
I've applied your changes and now is working! Thanks again for your work.

**Comment by benoitjchevalier** [comment]
@jhildenbiddle thanks for the quick turnaround and very detailed explanation, this also helped me a lot!

**Comment by jhildenbiddle** [comment]
@dvelasquez @benoitjchevalier -- Excellent! Glad the ponyfill is working for you.

I just published 1.11.1 to npm this evening (see [CHANGELOG.md](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/CHANGELOG.md) for details) which will be of interest to those working with web components. From the changelog:

> - Added logic to automatically get document-level CSS custom property values when option.rootElement has been set to a shadow host or root. These values are required by the polyfill to transform shadow <link> and <style> nodes. This new behavior makes it possible to target a shadow host or root element without manually getting the document-level custom property values first.

What this does is replace this (from my previous post):

> Or call the ponyfill using options.rootElement to target a specific component:
> 
> ```javascript
> // Pass 1: Get custom properties from standard DOM
> cssVars({
>   updateDOM: false,
>   onComplete(cssText, styleNode, cssVariables) {
>     // Pass 2: Process shadow DOM with custom properties from standard DOM
>     cssVars({
>       rootElement: shadowRoot,
>       shadowDOM  : true,
>       variables  : cssVariables
>     });
>   }
> });
> ```

with this:

```javascript
cssVars({
  rootElement: document.querySelector('my-custom-element').shadowRoot,
});
```

Much easier. 

I've updated the [`options.shadowDOM`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsshadowdom) section in the docs with more examples. The takeaway is that it is now much easier to have a custom element update itself (by setting `rootElement` to the element) as well any nested custom elements (by setting `shadowDOM` to `true`), making it possible to bundle the ponyfill with your component if you prefer.

Hope this helps, and let me know if you have any questions. Thx!
